### DISCUSSION TITLE
Changes in the new ramp design

### DISCUSSION THREAD

**Discussion by manuelma** [spine-tools/SpineOpt.jl#788] [Ideas]
# Introduction

The purpose of this post is to present the current SpineOpt's ramp implementation and then the changes introduced by the new design. The ultimate goal is to allow for a more clear appreciation of the benefits and drawbacks and also to help the development work.

@gnawin @datejada @g-moralesespana @mihlema @DillonJ @nnhjy 

# Common stuff

This stuff is required by the current and/or the new implementations (but none of them *introduces* it).

## Variables
| name | description |
| --- | --- |
| `unit_flow` | flow on a unit |
| `units_on` | number of units online |
| `units_started_up` | number of units started up |
| `units_shut_down` | number of units shut down |
| `nonspin_units_started_up` | number of units held for non-spinning upwards reserve provision |
| `nonspin_units_shut_down` | number of units held for non-spinning downwards reserve provision |

## Parameters
| name | description |
| --- | --- |
| `minimum_operating_point` | lower bound on the unit's flow |

# Current implementation

## Parameters
| name | description |
| --- | --- |
| `ramp_up_limit` | upper bound on the unit's ramp up flow |
| `ramp_down_limit` | upper bound on the unit's ramp down flow |
| `min_startup_ramp`, `max_startup_ramp` | lower and upper bound on the unit's startup flow |
| `min_shutdown_ramp`, `max_shutdown_ramp` | lower and upper bound on the unit's shutdown flow |
| `min_res_startup_ramp`, `max_res_startup_ramp` | lower and upper bound on the unit's non-spinning startup flow |
| `min_res_shutdown_ramp`, `max_res_shutdown_ramp` | lower and upper bound on the unit's non-spinning shutdown flow |
| `ramp_up_cost`, `ramp_down_cost` | costs associated to ramp up and down flows |

## Variables
| name | description |
| --- | --- |
| `ramp_up_unit_flow` | unit's ramp up flow |
| `ramp_down_unit_flow` | unit's ramp down flow |
| `start_up_unit_flow` | unit's startup flow |
| `shut_down_unit_flow` | unit's shutdown flow |
| `nonspin_ramp_up_unit_flow` | unit's non-spinning startup flow |
| `nonspin_ramp_down_unit_flow` | unit's non-spinning shutdown flow |

NOTE: These variables are not generated unless they are used.

### Ramp up and down flows within the limits
```
ramp_up_unit_flow <= (units_on - units_started_up) * ramp_up_limit

ramp_down_unit_flow <= (units_on - units_started_up - nonspin_units_shut_down) * ramp_down_limit
```
NOTE: It feels that we should also subtract `nonspin_units_shut_down` to `units_on` on the rhs of the first constraint
to make it more symmetrical to the second constraint, but I'm just guessing...

### Start up and shut down flows within the limits
```
units_started_up * min_startup_ramp <= start_up_unit_flow <= units_started_up * max_startup_ramp

units_shut_down * min_shutdown_ramp <= shut_down_unit_flow <= units_shut_down * max_shutdown_ramp
```

### Non-spinning ramp up and down flows within the limits
```
nonspin_units_started_up * min_res_startup_ramp <= nonspin_ramp_up_unit_flow <= nonspin_units_started_up * max_res_startup_ramp

nonspin_units_shut_down * min_res_shutdown_ramp <= nonspin_ramp_down_unit_flow <= nonspin_units_shut_down * max_res_shutdown_ramp
```

### Split ramps
```
+ unit_flow[t] - unit_flow[t-1]
+ unit_flow[t, upward_reserve] - unit_flow[t, downward_reserve]
==
+ ramp_up_unit_flow[t] + start_up_unit_flow[t] + nonspin_ramp_up_unit_flow[t]
- ramp_down_unit_flow[t] - shut_down_unit_flow[t] - nonspin_ramp_down_unit_flow[t]
```

## Objective

```
... + ramp_up_unit_flow * ramp_up_cost + ramp_down_unit_flow * ramp_down_cost + ...
```

# New design

## Parameters
| name | description |
| --- | --- |
| `ramp_up_limit` | upper bound on the unit's ramp up flow |
| `ramp_down_limit` | upper bound on the unit's ramp down flow |
| `startup_limit` | upper bound on the unit's startup flow (spinning or not) |
| `shutdown_limit` | upper bound on the unit's shutdown flow (spinning or not) |

## Variables
No new variables.

### Ramp up flow within the limits

```
+ unit_flow[t] - unit_flow[t-1] + unit_flow[t, upward_reserve]
<=
+ (startup_limit - minimum_operating_point - ramp_up_limit) * units_started_up[t]
+ (minimum_operating_point + ramp_up_limit) * units_on[t]
- minimum_operating_point * units_on[t-1]
```

### Ramp down flow within the limits

```
+ unit_flow[t-1] - unit_flow[t] + unit_flow[t, downward_reserve]
<=
+ (shutdown_limit - minimum_operating_point - ramp_down_limit) * units_shut_down[t]
+ (minimum_operating_point + ramp_down_limit) * units_on[t-1]
- minimum_operating_point * units_on[t]
```

## Objective
No new objective terms.

# Changes

1. Get rid of all the lower bound parameters, `min_startup_ramp`, `min_shutdown_ramp`, `min_res_startup_ramp`, `min_res_shutdown_ramp`; we just need `minimum_operating_point`.
2. Get rid of `max_res_startup_ramp` and `max_res_shutdown_ramp` parameters; instead, use `max_startup_ramp` and `max_shutdown_ramp` also for the non-spinning reserve flows - but rename these parameters to `startup_limit` and `shutdown_limit`.
3. Get rid of all the current ramp variables.
4. Get rid of all the current constraints and just replace them by the new constraints (not enough overlapping).
5. Get rid of the current ramp objective terms.

**Comment by manuelma** [comment]
Can anybody please confirm that the above is correct and we are all onboard with the changes? It seems like the new design trades some of the flexibility by simplicity - maybe that's good, I don't know. The lack of ramp costs in the new design is more concerning, maybe we want to introduce them? The difficulty then is we do not have the dedicated ramp up/down flow variables to write the objective function...

**Reply by gnawin** [reply]
Yes, with our proposal, you can still model the costs of start-up/shutdown because they are only related to start-up/shutdown variables (as it is also with the current objective terms), but ramp costs seem to be a bit of a far-stretch. But to me, it is not entirely clear why you would need these costs, as you have that covered already in costs related to flows/start-up/shutdown...
maybe I miss something, but can someone explain a bit here the rationale?
@manuelma @datejada @mihlema

**Reply by manuelma** [reply]
Right, startup and shutdown costs are already implemented in SpineOpt. They are of course completely unrelated to ramps. The new ramp design doesn't touch that at all.

**Reply by gnawin** [reply]
Thanks for confirming that. I was a bit confused by the rationale of using ramp objective terms in the first place (as it is with the current design). Are they needed?

**Reply by nnhjy** [reply]
In my understanding, ramp objective terms are there for any potential cases where ramping costs need to be represented. That is a kind of flexibility SpineOpt aims to provide.

**Reply by datejada** [reply]
@nnhjy still having a (ramping cost + starting/shutdown cost +  flow's variable cost) is redundant in the current implementation since the `start_up_unit_flow` and `shut_down_unit_flow` are accounted into the `unit_flow` variable (due to the splitting ramps constraint). So, you already have a variable cost in the flow variable. In addition, any fixed cost that you want to account for the start/shutdown ramping should be accounted for in the starting/shutdown cost. So, `start_up_unit_flow` and `shut_down_unit_flow` are just auxiliary variables in the formulation to easily write the current constraints; they don't have any economic meaning (unless I'm missing something here, but I can´t foresee any economic meaning for the ramping that can't be accounted in the other two concepts). That's why, I don't think having a ramping cost provides flexibility; on the contrary, I think it confuses the user (again, I might be overlooking the use case for that cost; if there is one, at least it would be good to have it mentioned in the documentation, why and when this is useful).

@gnawin any extra thoughts on this?

**Reply by manuelma** [reply]
Thank you @datejada - just a small clarification, the current objective term is `ramp_up_unit_flow * ramp_up_cost + ramp_down_unit_flow * ramp_down_cost` - so the variables of concern are `ramp_up_unit_flow` and `ramp_down_unit_flow` - not `start_up_unit_flow` and `shut_down_unit_flow`. But what you're saying is still valid in my opinion, and I feel it's okay to drop that objective term. @mihlema @DillonJ @jkiviluo @nhniina @Tasqu @tarskul  what do you think?

**Reply by datejada** [reply]
@manuelma I agree; those are the variables I meant in my previous comment. Thanks for the clarification

**Reply by nhniina** [reply]
Thanks for the description! This looks relatively similar to what is implemented for outputs of units with online variables, for exampe, in Backbone. Some questions: Does this work for unit inputs (and is it so that for them `unit_flow` < 0)? What are `startup_limit` and `shutdown_limit` actually? Are they like ramp rates when starting up / shutting down? What is the role of `minimum_operating_point` in the equations?

Backbone implements ramping costs as different tiers, and although they probably haven't been used that much, I think they are a nice feature. The way they were implemented was quite complicated, though, so maybe we want to leave them out first and add later.

**Reply by manuelma** [reply]
> Does this work for unit inputs (and is it so that for them unit_flow < 0)?

It does work for unit input flows, or in other words, flows from a node to a unit. Flows are always positive in SpineOpt though.

> What are startup_limit and shutdown_limit actually? Are they like ramp rates when starting up / shutting down? 

Yeah, I think you have the right idea. They limit the flow when the unit starts up and shuts down, respectively.

> What is the role of minimum_operating_point in the equations?

For example, for the first constraint in the new design, when the unit is shutting down, we have

`units_started_up[t] == units_on[t] == unit_flow[t] == 0` (because the unit just became offline), and
`units_on[t-1] == 1` (because the unit was online the previous step)

If we replace all that in the constraint (and ignore the reserves just for simplicity) we get

`- unit_flow[t-1] <= - minimum_operating_point`

which of course is equivalent to

`unit_flow[t-1] >= minimum_operating_point`

So basically the role of minimum_operating_point is to ensure the constraint remains valid when the unit is shutting down.

For the second constraint, the role is to ensure the constraint remains valid when the unit is starting up.

**Reply by nhniina** [reply]
Is the idea to still have a separate minimum operating point equation in SpineOpt? Does this formulation work with that if the unit is starting up and `startup_limit` is not high enough to reach minimum load in one time step?

And how is upward and downward reserve defined for electricity consumption units? Is it providing upward or downward reserve when it's increasing its consumption? From system perspective it would be downward reserve but it would also be ramping up from its own perspective at the same time. Or is it providing upward reserve and then it's converted to system downward reserve in the reserve requirement constraint?

(Sorry for being quite slow in figuring out all these directions in SpineOpt...)

**Reply by manuelma** [reply]
No worries, I think the idea is to keep the minimum operating point constraint because it does more stuff, but I'm not sure. 
I also don't know the answer to the reserves question either. @gnawin @datejada @g-moralesespana maybe you can help?

**Reply by manuelma** [reply]
> Does this formulation work with that if the unit is starting up and startup_limit is not high enough to reach minimum load in one time step?

About this, SpineOpt will check that start_up_limit and shut_down_limit are both greater than minimum_operating_point before running, and will issue an error describing the situation - should be fine.

**Reply by datejada** [reply]
Hi @nhniina I get your point, and I'm not quite sure about the answer. I think if is a consumption unit, then the relationship `unit_from__node` will have the ramping parameters. So the flow will go from the `node` to the `unit`, meaning that an increase in that flow (ramping up) is an increase in the consumption of the unit, meaning providing downward reserve. So, you need to be careful on how you set up your **downward reserve node group** because then you need to create the relationship using the `unit_from__node` instead of `unit__to_node` (which will be the normal one you would use for production units).

So, long story short, it should work fine, but you need to be careful with the set-up of your reserves and ramping in SpineOpt.

I think it might be a good candidate to the section *"How-To"* or *Tutorial*

Finally, I would like to highlight that this kind of setup for the consumption unit is also for the current formulation in SpineOpt, and that our proposal doesn't change anything in that regard.

**Reply by gnawin** [reply]
> Is the idea to still have a separate minimum operating point equation in SpineOpt? 

Hi @nhniina, 
yes, my understanding is that the minimum operating point constraint is still useful because it includes non-spinning reserves. Ramping constraints, although becoming a minimum operating constraint at some conditions, don't consider non-spinning reserves. 

Therefore, if there are no non-spinning reserves, it will be the same with the ramping constraints. Otherwise, it will overrule the ramping constraints.

**Reply by nhniina** [reply]
Thanks for all your answers! I perhaps missed something in the reserve explanation because I'm still confused why ramp up constraint lhs is not `+ unit_flow[t] - unit_flow[t-1] + unit_flow[t, downward_reserve]`. I'm also not sure about the units of measurement and whether the idea is to divide or multiply something with time step duration because normally ramp (and maybe shutdown/start-up) limits are given in MW/h or similar while the others are given in MW or similar.

**Reply by manuelma** [reply]
I skipped those multipliers in the description for brevity but they are there.

**Reply by datejada** [reply]
@nhniina I think I see your point for the first part, and I agree that for the consumption units, it is unclear if it will work or how it should be set up to work (I think even the current constraints can't handle it correctly). So, it is better to include the discussion in the next developers meeting to have a consensus on how to do it. In the meantime, I suggest continuing with the implementation as proposed and adding/changing it if needed for the consumption units. Is that okay?

**Reply by manuelma** [reply]
sounds good @datejada - it will be a small change anyways so no reason to pause implementation.

**Reply by g-moralesespana** [reply]
> @nhniina And how is upward and downward reserve defined for electricity consumption units? Is it providing upward or downward reserve when it's increasing its consumption? 

you are right on this point. The eqs are exactly the same for consumption or generating units. The difference is now that for consumption units the upwards (and downwards) reserve contributions become downwards (and upwards). So the implementation should ensure that.

**Reply by manuelma** [reply]
We need to clarify this point. So you're saying that e.g. in the ramp up constraint, the third term in the lhs sometimes needs to be unit_flow[t, downward_reserve]?

**Reply by g-moralesespana** [reply]
Yes, for consumption units it should be unit_flow[t, downward_reserve]. For production units, it should be unit_flow[t, upward_reserve].
The constraints are all exactly the same for production and consumption unit commitment units, but consumption units change the direction of everything: the production becomes consumption, the upward_reserve becomes downward_reserve, and the downward_reserve becomes upward_reserve (and the same with nonspinning reserves).

**Reply by manuelma** [reply]
I now realised that we should be talking about input and output flows rather than about consumption and production units - the same unit can be consuming and producing in the general case.

So @g-moralesespana can you perhaps explain the rule again but referring to input and output flows? My understanding is a unit can't be consuming reserves (just providing them), so when constraining the ramps on the input side, we shouldn't have any reserve flows in the equation. Is that correct?

**Reply by manuelma** [reply]
@g-moralesespana @datejada @gnawin this is the only detail that's missing I think (the effect of the flow direction, input or output, on the reserves flow term). As soon as we deal with this we can merge.

**Reply by manuelma** [reply]
Hi @g-moralesespana @datejada @gnawin 

Maybe you/we need more time to think but in the meantime I think the best would be to just merge https://github.com/spine-tools/SpineOpt.jl/pull/789 so I can start using the new parameter names already in the unit flow capacity constraint. Otherwise I would be doing double work. The current ramp implementation works if no reserves, the only detail is choosing upwards or downwards reserves depending on the flow direction and that should be easy to do once we have clarity.

(Unfortunately I don't think I am able to achieve that clarity on my own so please, whenever you get the chance, if you could answer my question above that would be great. Thank you!)

**Reply by nhniina** [reply]
When there is too much load in the system compared to generation, the system frequency starts to decrease and upward reserve is needed to return frequency to the nominal value. Upward reserve can be provided by generation units by increasing output and by consumption units by decreasing input. For example, an electric boiler can provide upward reserve by decreasing its electricity consumption (and provide downward reserve by increasing its electricity consumption).

In SpineOpt, when a unit is decreasing its input, it is ramping down in that node (if I understood correctly the explanations above). So, whenever a unit is providing upward reserve to its input node, it needs to be taken into account in the ramp down constraint.

I believe there are 4 combinations we need to consider

- unit output, ramp up constraint, upward reserve provision
- unit output, ramp down constraint, downward reserve provision
- unit input, ramp up constraint, downward reserve provision
- unit input, ramp down constraint, upward reserve provision

**Reply by datejada** [reply]
@g-moralesespana, @gnawin, and I will discuss today how to cover the last two cases (the two first ones are covered with the current formulation. We want to double-check the RHS of the constraints for these cases, and then we will get back with a proposal and discuss how it can be implemented in SpineOpt.

**Reply by manuelma** [reply]
Thank you very much @nhniina for the explanation. And thank you @datejada for the update, sounds like a plan, very good.

**Reply by g-moralesespana** [reply]
To generalize the reply of @nhniina, we need to consider more constraints affecting the 4 cases:

- unit output, ramp up & upper bound unit capacity constraints, upward reserve provision
- unit output, ramp down & lower bound unit capacity constraints, downward reserve provision
- unit input, ramp up & upper bound unit capacity constraints, downward reserve provision
- unit input, ramp down & lower bound unit capacity constraints, upward reserve provision

and the logic for nonspinning reserve:
- unit output, uptime & upper bound constraints, nonspinning_unit_shutdown
- unit output, downtime constraints, nonspinning_unit_startup
- unit input, uptime & upper bound constraints, nonspinning_unit_startup
- unit input, downtime constraints, nonspinning_unit_shutdown

**Reply by manuelma** [reply]
Sounds good @g-moralesespana - just to understand, there will be changes to more constraints than just the ramp constraints? It would be good to specify those changes with respect to the overleaf document https://www.overleaf.com/project/6454fde7fdf7f1c73ceeb2a0 - maybe highlight the new terms in red or something. I can work on that in a new PR.

**Reply by datejada** [reply]
We continue the discussion for consumption units providing reserves here: https://github.com/spine-tools/SpineOpt.jl/discussions/796

**Comment by nnhjy** [comment]
So the following variables will remain in the new implementation concerning that:
> No new variables.

and

> 3. Get rid of all the current ramp variables.

Is that correct?

## Common Variables
| name | description |
| --- | --- |
| `unit_flow` | flow on a unit |
| `units_on` | number of units online |
| `units_started_up` | number of units started up |
| `units_shut_down` | number of units shut down |
| `nonspin_units_started_up` | number of units held for non-spinning upwards reserve provision |
| `nonspin_units_shut_down` | number of units held for non-spinning downwards reserve provision |

## Specific Variables
| name | description |
| --- | --- |
| `start_up_unit_flow` | unit's startup flow |
| `shut_down_unit_flow` | unit's shutdown flow |

**Reply by manuelma** [reply]
No, sorry, `start_up_unit_flow` and `shut_down_unit_flow` will also be gone - I included them in the 'current ramp variables' because they only serve the current ramp implementation and nothing more.

**Comment by manuelma** [comment]
Since nobody seems to complain I think it's time for us to finally try and implement these changes. If you are ok @gnawin @datejada I can take the first stab, seems pretty straightforward since it's mostly getting rid of old stuff and adding new stuff (no code transformation so to say.)

I can work on a new branch and publish as soon as possible. Maybe I can write the code and you @gnawin can write the tests, does that sound ok?

**Reply by gnawin** [reply]
Sure. Thank you very much for the work @manuelma. Once the constraint is finished, I will take a shot at the tests.

**Reply by manuelma** [reply]
Hi @gnawin 

I opened a draft PR with the ramp changes https://github.com/spine-tools/SpineOpt.jl/pull/789

My part is basically done, all tests pass except the ones for the ramp up and down constraints, which are the ones I was hoping you could fix (but I could do it too if you are short on time)...

**Comment by gnawin** [comment]
> Hi @gnawin
> 
> I opened a draft PR with the ramp changes #789
> 
> My part is basically done, all tests pass except the ones for the ramp up and down constraints, which are the ones I was hoping you could fix (but I could do it too if you are short on time)...

Hi @manuelma,

It was my initial idea to write this test as well. But since your part is already done, and I have not started fully on the other test (for the unit capacity), time-wise, it will be most favorable if you can do this now. 

Meanwhile, I will try my best on the other one, but it may be not ideal in case it takes too long (either because I'm stuck somewhere and/or I'm short of time) and you're already on top of this. Will keep you closely posted on my progress, and say if I can't finish it this week I'll hand it to you. Does that sound okay?

**Comment by DillonJ** [comment]
Hi folks,

Not sure if this is the right place to comment - but we just had a discussion on ramp rates and there was an issue where the actual ramp rate in the old formulation depended on the temporal resolution of the unit_flow variable. That is one issue. A second issue is that if the temporal changes over time, as it can and as we support, then the effective ramp rate changes too which would not be expected behaviour.

For example - we have models that have a low resolution look-ahead for long term storage and if a ramp rate was imposed, this would likely lead to infeasibilities. I think we need to make sure of three things:
1. The effective ramp rate shoud take account of the duration of timeslices and account for changes in temporal resolution correctly
2. The name of the ramping parameters should be explicit regarding the assumed temporal units. This is an exception in SpineOpt where data usually defines units - but in cases regarding duration, we need to be explicit.
3. The naming of the parameters should make it clear that the value is relative to the unit_flow capacity. E.g. `relative_ramp_up_limit_per_hour` (or something like that).

**Reply by manuelma** [reply]
Good points @DillonJ 

My impression is we handle the above by multiplying relevant terms in the constraint by the duration of the corresponding time slice. That means the ramp value is relative to the `model`'s `duration_unit` which to me makes a lot of sense, but we need to write it down in the documentation of all the ramp parameters. In other words, clarify that they are both relative to the capacity and to the duration unit of the model.

I will write unit-tests too to verify that the behavior meets the above points.

Regarding naming, we have `ramp_up_limit` and `ramp_down_limit` which I think might work. Personally I'd prefer to clarify the meaning in the documentation rather than having a longer self-explanatory name, but I'm open.

**Reply by datejada** [reply]
Thank you @manuelma, I agree with your answer to @DillonJ

**Reply by jkiviluo** [reply]
Right, the unit should be p.u. / min (or another time unit).

While I think the names should be descriptive, I also think it's enough in this case to have `ramp_up_limit` and then start the parameter description with the unit. `relative` is often difficult to interpret - it's not obvious what the thing is relative to.

**Reply by DillonJ** [reply]
The main thing is that it is clear, at least in the documentation, about how the parameter is used and how it is interpreted from a temporal resolution point of view - so the implementation is sensible in that regard.

I think it's probably fine that we make it relative to the model duration unit - but a user might not be familiar with that or its implications so in the documentation relating to ramp_up/down_limit, we should say what the default is as in the vast majority of cases, a user won't have changed it.

**Reply by g-moralesespana** [reply]
@DillonJ and @manuelma I agree that the ramp should be multiplied by the duration of the current time slice, which also means it can change per time slice if the duration of the time slice changes.

**Comment by manuelma** [comment]
I have one more question here @gnawin @datejada @g-moralesespana 

The new contraints do not consider the variables `nonspin_units_started_up` and `nonspin_units_shut_down` - is that intended? Don't we need to consider non-spinning reserve provision as well in the context of start_up_limit/shut_down_limit?

**Reply by datejada** [reply]
Hi @manuelma, We discussed it with @g-moralesespana, and it is not necessary to have them, but it will be nice to add them. We are reviewing the constraints to include them so they are still the tightest possible. Once it is done, we can add the terms to the constraints you are implementing. Is that ok?

**Reply by manuelma** [reply]
Ah nice, thank you very much @datejada

**Reply by g-moralesespana** [reply]
Hi @manuelma, thanks for pointing this out. Indeed a tighter constraint for ramp down limit is:
![image](https://github.com/spine-tools/SpineOpt.jl/assets/42405171/2801e0d9-98f6-43cf-9c95-415aae02a853)

**Reply by manuelma** [reply]
Nice one @g-moralesespana - and we can't add u_{t-1}^{N+} to y_t in the ramp up constraint? That would make everything symmetric and more awesome... but maybe it just doesn't work?

**Reply by g-moralesespana** [reply]
@manuelma I agree with the part that it will make it more awesome and symmetric, and I also agree with the part that it would not work:  since u_{t-1}^{N+} does not constrain the production of the unit p_t, unlike u_{t-1}^{N-} which ensures the unit should be able to turn off to provide nonspinning towards reserves (meaning p_t <= SD).

**Reply by g-moralesespana** [reply]
@manuelma I just realized that adding u_{t-1}^{N-} to the ramp down changes the intended ramp constraint behaviour. Actually, it should not be added at all. So following the suggestion of @manuelma to leave the ramp constraints symmetric and more awesome, the ramp constraints should stay how they were:
![image](https://github.com/spine-tools/SpineOpt.jl/assets/42405171/ed36ba95-da46-467f-be48-224a36d1f6dd)
sorry for the confusion.

**Reply by g-moralesespana** [reply]
> @manuelma Don't we need to consider non-spinning reserve provision as well in the context of start_up_limit/shut_down_limit?

Yes, we need to. u_{t-1}^{N-} already appears in the context of shutdown_limit:
![image](https://github.com/spine-tools/SpineOpt.jl/assets/42405171/f381fa13-77ac-4d6b-95cf-6987d74969e9)
![image](https://github.com/spine-tools/SpineOpt.jl/assets/42405171/94fd26e1-276f-4cac-aefd-08f7d151b3be)
these constraints ensure that the unit should be producing below its shutdown_limit to be able to provide nonspinning downwards reserve, which makes sense in the way that the unit should be able to shutdown.

And the complementary constraints guaranteeing that the units' startup_limit/shutdown_limit is the maximum for nonspinning reserves are
![image](https://github.com/spine-tools/SpineOpt.jl/assets/42405171/f5446175-8fc2-47a9-b864-f4ca17236a73)

**Reply by manuelma** [reply]
Thank you @g-moralesespana - so it is handled in other constraints.

**Reply by jkiviluo** [reply]
Below is a comment of mine in a previous discussion about reserve formulations. I think it's still relevant. Although maybe doesn't need to impact what's going on in this issue at this stage, but maybe just a further evolution - I think it's impact would be mostly on the data side. Including dropping the name 'non-spinning', because it's inadequate.

-	I don’t particularly like the distinction between spinning and non-spinning. It should be about reaction time. Sometimes that reaction time is limited by start-ups and sometimes you can have a reaction time even though the unit is ‘online’. This could be due to e.g. communication delay or valves. The reaction time can prevent some resources from contributing to all reserve categories. It all can probably be represented on the data side before applying the equations.

**Reply by manuelma** [reply]
Thank you @jkiviluo that's interesting. So you suggest dropping those two variables altogether?
Anyways that discussion belongs to the reserves I believe. My plan is to open a new discussion similar to this one but about the reserves so we can look at those details.

**Reply by g-moralesespana** [reply]
I completely agree with @jkiviluo, the reserves are products that are demanded in different time frames, it does not matter that much if they are spinning and not spinning. I think we can move on with this topic. And as @manuelma proposed, it is better to open a new discussion about this. Anyway, the formulation is (almost) the same but some names should change.

**Reply by manuelma** [reply]
Here is the discussion about reserves https://github.com/spine-tools/SpineOpt.jl/discussions/795

Although, I feel that moving away from spinning/non-spinning might be better done in a second round. I suggest we first get it right with spinning/non-spinning and then we generalize into 'speed' - otherwise the change might become too overwhelming.

**Comment by manuelma** [comment]
> Hi @manuelma, thanks for pointing this out. Indeed a tighter constraint for ramp down limit is: ![image](https://user-images.githubusercontent.com/42405171/281408334-2801e0d9-98f6-43cf-9c95-415aae02a853.png)

Quick question @g-moralesespana - the term in red above is a mistake, right? You corrected it in a comment in this discussion but the overleaf (where the equation is numbered as 11) still has it, see here https://www.overleaf.com/project/6454fde7fdf7f1c73ceeb2a0

**Reply by g-moralesespana** [reply]
Yes, it is a mistake. I just corrected the overleaf. Sorry again for the confusion.
